cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -0,0 +1,172 @@
+
+	"github.com/spf13/cobra"
+
+	//clientset "k8s.io/client-go/kubernetes"
+	// kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
+	// "k8s.io/api/core/v1"
+	// extensions "k8s.io/api/extensions/v1beta1"
+	// metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	// "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	//"k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/selfhosting"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/upgrade"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/version"
+)
+
+// applyFlags holds the information about the flags that can be passed to apply
+type applyFlags struct {
+	flagYes, force, dryRun bool
+	newK8sVersionStr       string
+	newK8sVersion          *version.Version
+	parent                 *cmdUpgradeFlags
+}
+
+// NewCmdApply returns the cobra command for `kubeadm upgrade apply`
+func NewCmdApply(parentFlags *cmdUpgradeFlags) *cobra.Command {
+	flags := &applyFlags{
+		parent: parentFlags,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "apply",
+		Short: "Apply the upgrade",
+		Run: func(cmd *cobra.Command, args []string) {
+			kubeadmutil.CheckErr(ValidateFlags(flags))
+
+			err := RunApply(flags)
+			kubeadmutil.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().BoolVar(&flags.flagYes, "yes", flags.flagYes, "Actually perform the selected upgrade")
+	cmd.Flags().BoolVar(&flags.force, "force", flags.force, "Force upgrading although some requirements might not be met")
+	cmd.Flags().BoolVar(&flags.dryRun, "dry-run", flags.dryRun, "Do not change any state, just output what actions would be applied")
+	cmd.Flags().StringVar(&flags.newK8sVersionStr, "version", flags.newK8sVersionStr, "Selected version to upgrade to")
+
+	return cmd
+}
+
+// RunApply takes care of the actual upgrade functionality
+func RunApply(flags *applyFlags) error {
+
+	// Start with the basics, verify that the cluster is healthy and get config
+	client, cfg, err := EnforceRequirements(flags.parent.kubeConfigPath, flags.parent.cfgPath, flags.parent.printConfig)
+	if err != nil {
+		return err
+	}
+
+	// Set the version on the config object now
+	cfg.KubernetesVersion = flags.newK8sVersionStr
+
+	internalcfg := &kubeadmapi.MasterConfiguration{}
+	api.Scheme.Convert(cfg, internalcfg, nil)
+
+	fmt.Printf("[upgrade/version] You have chosen to upgrade to version %q\n", flags.newK8sVersionStr)
+	versionGetterImpl := upgrade.NewKubeVersionGetter(client, os.Stdout)
+	versionSkewErrs := upgrade.EnforceVersionPolicies(versionGetterImpl, flags.newK8sVersionStr, flags.newK8sVersion, flags.parent.allowExperimentalUpgrades, flags.parent.allowRCUpgrades)
+	if versionSkewErrs != nil {
+
+		if len(versionSkewErrs.Mandatory) > 0 {
+			fmt.Printf("[upgrade/version] FATAL: The --version argument is invalid: %v.\n[upgrade/version] Can be bypassed if you pass the --force flag\n", versionSkewErrs.Mandatory)
```

Comment:
Is this correct? I thought mandatory errors could not be bypassed with `--force`?